### PR TITLE
fix: move offline/cache/interception switches to BrowserContext

### PR DESCRIPTION
**Pull request by yury-s**

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -212,6 +212,9 @@ Indicates that the browser is connected.
     - `deviceScaleFactor` <[number]> Specify device scale factor (can be thought of as dpr). Defaults to `1`.
     - `isMobile` <[boolean]> Whether the `meta viewport` tag is taken into account. Defaults to `false`.
   - `userAgent` <?[string]> Specific user agent to use in this context.
+  - `cacheEnabled` <?[boolean]> Toggles page cache in the context. By defaul caching is enabled.
```
Comment (dgozman) [inline docs/api.md]: page cache -> http cache

File: docs/api.md [hunk 2]
```diff
@@ -321,6 +327,12 @@ Creates a new page in the browser context and optionally navigates it to the spe
 
 An array of all pages inside the browser context.
 
+#### browserContext.setCacheEnabled([enabled])
+- `enabled` <[boolean]> sets the `enabled` state of the cache.
```
Comment (dgozman) [inline docs/api.md]: "http cache"

File: docs/api.md [hunk 3]
```diff
@@ -353,6 +365,10 @@ await browserContext.setGeolocation({latitude: 59.95, longitude: 30.31667});
 
 > **NOTE** Consider using [browserContext.setPermissions](#browsercontextsetpermissions-permissions) to grant permissions for the page to read its geolocation.
 
+#### browserContext.setOfflineMode(enabled)
+- `enabled` <[boolean]> When `true`, enables offline mode for the page.
```
Comment (dgozman) [inline docs/api.md]: page -> context

File: docs/api.md [hunk 4]
```diff
@@ -380,6 +396,32 @@ const context = browser.defaultContext();
 await context.setPermissions('https://html5demos.com', ['geolocation']);
 ```
 
+#### browsercontext.setRequestInterception(enabled)
+- `enabled` <[boolean]> Whether to enable request interception.
+- returns: <[Promise]>
+
+Activating request interception enables `request.abort`, `request.continue` and
+`request.respond` methods.  This provides the capability to modify network requests that are made by a page.
```
Comment (dgozman) [inline docs/api.md]: "by all pages in the context"

File: docs/api.md [hunk 5]
```diff
@@ -380,6 +396,32 @@ const context = browser.defaultContext();
 await context.setPermissions('https://html5demos.com', ['geolocation']);
 ```
 
+#### browsercontext.setRequestInterception(enabled)
```
Comment (dgozman) [inline docs/api.md]: typo: browser_C_ontext

File: docs/api.md [hunk 6]
```diff
@@ -380,6 +396,32 @@ const context = browser.defaultContext();
...
+    interceptedRequest.abort();
+  else
+    interceptedRequest.continue();
+});
+await context.setRequestInterception(true);
+await page.goto('https://example.com');
+await browser.close();
+```
+
+> **NOTE** Enabling request interception disables page caching.
```
Comment (dgozman) [inline docs/api.md]: page caching -> http cache.

File: src/chromium/crNetworkManager.ts [hunk 7]
```diff
@@ -91,17 +97,19 @@ export class CRNetworkManager {
   }
 
   async setCacheEnabled(enabled: boolean) {
```
Comment (dgozman) [inline src/chromium/crNetworkManager.ts]: Why do we have a parameter which is now ignored?

File: src/firefox/ffPage.ts [hunk 8]
```diff
@@ -90,6 +90,8 @@ export class FFPage implements PageDelegate {
       promises.push(this._session.send('Page.setJavascriptEnabled', { enabled: false }));
     if (options.userAgent)
       promises.push(this._session.send('Page.setUserAgent', { userAgent: options.userAgent }));
+    if (options.cacheEnabled === false)
```
Comment (dgozman) [inline src/firefox/ffPage.ts]: It seems like this is missing request interception changes in firefox.

File: test/interception.spec.js [hunk 9]
```diff
@@ -634,8 +643,6 @@ module.exports.describe = function({testRunner, expect, defaultBrowserOptions, p
   describe('ignoreHTTPSErrors', function() {
     it('should work with request interception', async({newPage, httpsServer}) => {
       const page = await newPage({ ignoreHTTPSErrors: true, interceptNetwork: true });
-
-      await page.setRequestInterception(true);
```
Comment (dgozman) [inline test/interception.spec.js]: Why this change?

File: src/chromium/crNetworkManager.ts [hunk 10]
```diff
@@ -91,17 +97,19 @@ export class CRNetworkManager {
...
     await this._updateProtocolRequestInterception();
   }
 
-  async _updateProtocolRequestInterception() {
-    const enabled = this._userRequestInterceptionEnabled || !!this._credentials;
+  private _userRequestInterceptionEnabled() : boolean {
+    return !!this._page.browserContext()._options.interceptNetwork;
+  }
+
+  private async _updateProtocolRequestInterception() {
```
Comment (dgozman) [inline src/chromium/crNetworkManager.ts]: This stuff is too complicated 😞

File: docs/api.md (hunk 1 above)
Comment (yury-s) [reply docs/api.md]: Done.

File: docs/api.md (hunk 3 above)
Comment (yury-s) [reply docs/api.md]: Done.

File: docs/api.md (hunk 2 above)
Comment (yury-s) [reply docs/api.md]: Done.

File: src/chromium/crNetworkManager.ts (hunk 7 above)
Comment (yury-s) [reply src/chromium/crNetworkManager.ts]: This is how it is declared in the delegate interface. 2 other browsers use that param. We can update that in a separate change.

File: test/interception.spec.js (hunk 9 above)
Comment (yury-s) [reply test/interception.spec.js]: This is redundant. `interceptNetwork: true` above has the same effect.

File: src/chromium/crNetworkManager.ts (hunk 10 above)
Comment (yury-s) [reply src/chromium/crNetworkManager.ts]: it is, let me change that in a follow-up.

**Review by dgozman — APPROVED** [review]
